File: [0152][Medium][Maximum_Product_Subarray]/Maximum_Product_Subarray_2.java
// leetcode time     cost : 2 ms
// leetcode memory   cost : 38.4 MB
// solution 1, DP, Save Min value and Max value at each node and the result works out.
import java.math.*;
class Solution {
    public int maxProduct(int[] nums) {
        if (nums.length == 0) return 0;
        
        int res = nums[0];
        int positive = 1;
        int negative = 1;        
        for (int i = 0; i < nums.length; i++) {
            int x = nums[i];
            if (x >= 0) {
                positive = Math.max(positive * x, x);
                negative = negative * x;
            } else {
                int tmp = negative;
                negative = Math.min(positive * x, x);
                positive = tmp * x;
            }
            res = Math.max(res,positive);
        }
        return res;
    }
}

public class Maximum_Product_Subarray_2 {
    public static void main(String args[]) {
        int[] nums = {2,3,-2,4} ;   // expect is 6
        
        Solution Solution_obj = new Solution();
        int result = Solution_obj.maxProduct(nums);
        System.out.println("In java code,return value is :" + (result));
    }
}

File: [0152][Medium][Maximum_Product_Subarray]/Maximum_Product_Subarray_2.py
# -*- coding: utf-8 -*-  
# leetcode time     cost : 44 ms
# leetcode memory   cost : 13.7 MB
# solution 1, DP by scan forward and backward.
class Solution(object):
    def maxProduct(self, nums):
        res = -float('inf')
        product = 1
        length = len(nums)
        for i in range(length):
            product *= nums[i]
            res = max(product, res)
            if nums[i] == 0:
                product = 1
        # scan backward
        product = 1
        for j in range(length - 1, -1, -1):
            product *= nums[j]
            res = max(product, res)
            if nums[j] == 0:
                product = 1
        return res
    
    def maxProduct_2(self, nums):
        nums_rev = nums[::-1]
        for i in range(1, len(nums)):
            nums[i] *= nums[i - 1] or 1
            nums_rev[i] *= nums_rev[i - 1] or 1
        return max(nums + nums_rev)

def main():
    nums = [2,3,-2,4]        #expect is 6
    obj = Solution()
    result = obj.maxProduct(nums)
    print("return result is :",result)
    
if __name__ =='__main__':
    main()   